pkg/policyviolation/clusterpv.go

```diff
@@ -9,144 +9,103 @@ import (
+	cpvLister kyvernolister.ClusterPolicyViolationLister
+	// policy violation interface
+	kyvernoInterface kyvernov1.KyvernoV1Interface
 }
 
-func (gen *Generator) createCusterPV(info Info) error {
-	var pvs []kyverno.ClusterPolicyViolation
-	if !info.Blocked {
-		pvs = append(pvs, buildPV(info))
-	} else {
-		// blocked
-		// get owners
-		pvs = buildPVWithOwners(gen.dclient, info)
-	}
-	// create policy violation
-	if err := createPVS(gen.dclient, pvs, gen.pvLister, gen.pvInterface); err != nil {
-		return err
+func newClusterPV(dclient *client.Client,
+	cpvLister kyvernolister.ClusterPolicyViolationLister,
+	kyvernoInterface kyvernov1.KyvernoV1Interface,
+) *clusterPV {
+	cpv := clusterPV{
+		dclient:          dclient,
+		cpvLister:        cpvLister,
+		kyvernoInterface: kyvernoInterface,
 	}
-
-	glog.V(3).Infof("Created cluster policy violation policy=%s, resource=%s/%s/%s",
-		info.PolicyName, info.Resource.GetKind(), info.Resource.GetNamespace(), info.Resource.GetName())
-	return nil
-
+	return &cpv
 }
-func createPVNew(dclient *client.Client, pv kyverno.ClusterPolicyViolation, pvLister kyvernolister.ClusterPolicyViolationLister, pvInterface kyvernov1.KyvernoV1Interface) error {
-	var err error
+
+func (cpv *clusterPV) create(pv kyverno.PolicyViolation) error {
+	newPv := kyverno.ClusterPolicyViolation(pv)
 	// PV already exists
-	ePV, err := getExistingPVIfAny(pvLister, pv)
+	oldPv, err := cpv.getExisting(newPv)
 	if err != nil {
-		glog.Error(err)
-		return fmt.Errorf("failed to get existing pv on resource '%s' in namespace %s: %v", pv.Spec.ResourceSpec.ToKey(), pv.Namespace, err)
-	}
-	if ePV == nil {
-		// Create a New PV
-		glog.V(4).Infof("creating new policy violation for policy %s & resource %s/%s", pv.Spec.Policy, pv.Spec.ResourceSpec.Kind, pv.Spec.ResourceSpec.Name)
-		err := retryGetResource(pv.Namespace, dclient, pv.Spec.ResourceSpec)
-		if err != nil {
-			return fmt.Errorf("failed to retry getting resource for policy violation %s/%s: %v", pv.Name, pv.Spec.Policy, err)
-		}
-
-		_, err = pvInterface.ClusterPolicyViolations().Create(&pv)
-		if err != nil {
-			glog.Error(err)
-			return fmt.Errorf("failed to create cluster policy violation: %v", err)
-		}
-		glog.Infof("policy violation created for resource %v", pv.Spec.ResourceSpec)
-		return nil
+		return err
 	}
-	// Update existing PV if there any changes
-	if reflect.DeepEqual(pv.Spec, ePV.Spec) {
-		glog.V(4).Infof("policy violation spec %v did not change so not updating it", pv.Spec)
-		return nil
+	if oldPv == nil {
+		// create a new policy violation
+		return cpv.createPV(&newPv)
 	}
+	// policy violation exists
+	// skip if there is not change, else update the violation
+	return cpv.updatePV(&newPv, oldPv)
+}
 
-	pv.SetName(ePV.Name)
-	_, err = pvInterface.ClusterPolicyViolations().Update(&pv)
+func (cpv *clusterPV) getExisting(newPv kyverno.ClusterPolicyViolation) (*kyverno.ClusterPolicyViolation, error) {
+	pvs, err := cpv.cpvLister.List(labels.Everything())
```

Comment:
Can we list by label selector?

Reply:
Initial i removed this as labels were dynamic in the previous implementation and we handle updates or if labels are removed.
But with new updates, we are hardcoding labels so not labels don't give many advantages.

But i'm adding using label selectors